beacon-chain/sync/rpc_data_column_sidecars_by_range.go

```diff
@@ -0,0 +1,215 @@
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return errors.Wrap(err, "new block range batcher")
+	}
+
+	// Derive the wanted columns for the request.
+	wantedColumns := make([]uint64, len(request.Columns))
+	copy(wantedColumns, request.Columns)
+
+	// Sort the wanted columns.
+	slices.Sort(wantedColumns)
+
+	var batch blockBatch
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
+		batchStart := time.Now()
+		maxRequestDataColumnSidecars, err = s.streamDataColumnBatch(ctx, batch, maxRequestDataColumnSidecars, wantedColumns, stream)
+		rpcDataColumnsByRangeResponseLatency.Observe(float64(time.Since(batchStart).Milliseconds()))
+		if err != nil {
+			return err
+		}
+
+		// Once the quota is reached, we're done serving the request.
+		if maxRequestDataColumnSidecars == 0 {
+			log.WithField("initialQuota", beaconConfig.MaxRequestDataColumnSidecars).Debug("Reached quota for data column sidecars by range request")
+			break
+		}
+	}
+
+	if err := batch.error(); err != nil {
+		log.WithError(err).Debug("error in DataColumnSidecarsByRange batch")
+
+		// If we hit a rate limit, the error response has already been written, and the stream is already closed.
+		if !errors.Is(err, p2ptypes.ErrRateLimited) {
+			s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		}
+
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	closeStream(stream, log)
+	return nil
+}
+
+func (s *Service) streamDataColumnBatch(ctx context.Context, batch blockBatch, quota uint64, wantedDataColumnIndices []uint64, stream libp2pcore.Stream) (uint64, error) {
+	_, span := trace.StartSpan(ctx, "sync.streamDataColumnBatch")
+	defer span.End()
+
+	// Defensive check to guard against underflow.
+	if quota == 0 {
+		return 0, nil
+	}
+
+	// Loop over the blocks in the batch.
+	for _, block := range batch.canonical() {
+		// Get the block blockRoot.
+		blockRoot := block.Root()
+
+		// Retrieve the data column sidecars from the store.
+		verifiedRODataColumns, err := s.cfg.dataColumnStorage.Get(blockRoot, wantedDataColumnIndices)
+		if err != nil {
+			s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+			return quota, errors.Wrapf(err, "get data column sidecars: block root %#x", blockRoot)
+		}
+
+		// Write the retrieved sidecars to the stream.
+		for _, verifiedRODataColumn := range verifiedRODataColumns {
+			sidecar := verifiedRODataColumn.DataColumnSidecar
+			SetStreamWriteDeadline(stream, defaultWriteDuration)
+
+			if err := WriteDataColumnSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sidecar); err != nil {
+				s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+				tracing.AnnotateError(span, err)
+				return quota, errors.Wrap(err, "write data column sidecar chunk")
+			}
+
+			s.rateLimiter.add(stream, 1)
```

Comment:
there was a constant for rate limited amount above but this one is hard coded

Reply:
Fixed in 7e3a85562586cc2a295d813427085b6102bdb84b.